### ISSUE TITLE
PHP7 'Float' special class.

### ISSUE THREAD

**Issue by markkrueg** [LegendOfMCPE/EssentialsPE#190]
Running from current source (v2.0.0).
This occurs during boot up when running under PHP7. 
```
Error: Cannot use pocketmine\nbt\tag\Float as Float because 'Float' is a special class name
File: /EssentialsPE_v2.0.0.phar/src/EssentialsPE/BaseFiles/BaseAPI
Line: 40
Type: E_COMPILE_ERROR
```
Is there any workaround for this? I'd love to get EssentialsPE working again. 
```
[02:03:53] [Server thread/CRITICAL]: Error: "Call to a member function __destruct() on null" (EXCEPTION) in "/EssentialsPE_v2.0.0.phar/src/EssentialsPE/Loader" at line 110
[02:03:53] [Server thread/DEBUG]: #0 /src/pocketmine/plugin/PharPluginLoader(138): pocketmine\plugin\PluginBase->setEnabled(boolean )
[02:03:53] [Server thread/DEBUG]: #1 /src/pocketmine/plugin/PluginManager(636): pocketmine\plugin\PharPluginLoader->disablePlugin(EssentialsPE\Loader object)
[02:03:53] [Server thread/DEBUG]: #2 /src/pocketmine/plugin/PluginManager(626): pocketmine\plugin\PluginManager->disablePlugin(EssentialsPE\Loader object)
[02:03:53] [Server thread/DEBUG]: #3 /src/pocketmine/Server(2372): pocketmine\plugin\PluginManager->disablePlugins(boolean)
[02:03:53] [Server thread/DEBUG]: #4 /src/pocketmine/Server(2566): pocketmine\Server->forceShutdown(boolean)
[02:03:53] [Server thread/DEBUG]: #5 (): pocketmine\Server->crashDump(boolean)
```

**Comment by markkrueg** [comment]
It looks like the PHP7 compatible branches of PocketMine use the following:

```
use pocketmine\nbt\tag\ByteTag;
use pocketmine\nbt\tag\CompoundTag;
use pocketmine\nbt\tag\DoubleTag;
use pocketmine\nbt\tag\ListTag;
use pocketmine\nbt\tag\FloatTag;
use pocketmine\nbt\tag\IntTag;
use pocketmine\nbt\tag\LongTag;
use pocketmine\nbt\tag\ShortTag;
use pocketmine\nbt\tag\StringTag;
```

So I guess the idea would be to go through EssentialsPE and replace  pocketmine\nbt\tag\Float with  pocketmine\nbt\tag\FloatTag. Unfortunately I would guess this will make it incompatible with the PHP5 branch of PocketMine. 

I guess I will do this on my local copy of EssentialsPE; but maybe you want to make a branch for PHP7 or something?

**Comment by markkrueg** [comment]
I tried just changing line 40 from:
```
pocketmine\nbt\tag\Float
```
to
```
pocketmine\nbt\tag\FloatTag
```
And it appears to be working with PocketMine under PHP7! I haven't done extensive testing yet; but I tested several commands successfully.

**Comment by PEMapModder** [comment]
> I tried just changing line 40 from:
> ```
> pocketmine\nbt\tag\Float
> ```
> to
> ```
> pocketmine\nbt\tag\FloatTag
> ```
> 

This issue is caused by a backwards-incompatible change in PHP 7, as [described in PHP 7 migration notes](http://php.net/manual/en/migration70.incompatible.php#migration70.incompatible.other.classes):

> Invalid class, interface and trait names
> 
> The following names cannot be used to name classes, interfaces or traits:
> 
> * `bool`
> * `int`
> * `float`
> * `string`
> * `NULL`
> * `TRUE`
> * `FALSE`

We hence have to search all occurrences in the plugin where `pocketmine\nbt\tag\Int`, ``pocketmine\nbt\tag\String` and `pocketmine\nbt\tag\Float` class references are used directly (in `use` statements) or indirectly (_imported_ through `use` statements and the alias is used in the class body) and replace them.

However, we have missed an important change - PocketMine changed all NBT tags to have the suffix `Tag` in their class names, as of https://github.com/PocketMine/PocketMine-MP/commit/7f8b39a63c7d3d0849b722412f2cdaf362d80349, so we have to refactor them all.

> but maybe you want to make a branch for PHP7 or something?

:arrow_up: agreed.

@iksaku comments?

**Comment by markkrueg** [comment]
Suggestion: Maybe the main branch should really be the "PHP7 branch" and there should be a new branch created for PHP5. 

Genisys has done this with their fork of PocketMine-MP; and I'm guessing Shoghi would have done the same by now if he weren't busy. PHP7 is ready; and PHP5 is the past.

Whatever you wish. Huge thanks for continued development of EssentialsPE !

**Comment by PEMapModder** [comment]
As PocketMine hasn't officially adopted PHP 7 yet, EssentialsPE will support PHP 7 later.

**Comment by iksaku** [comment]
Right now, I will focus on finishing the 2.0.0 API changes, after that I will jump right into a `PHP7` branch, which will change a lot the coding style of the plugin, as well as fixing its namespaces.

**Comment by TheEpicCooldeal** [comment]
When will this be fixed. My main plugin is this....

**Comment by iksaku** [comment]
Soon...

**Comment by Tolo0** [comment]
I'm using a php7 build and just changed the tags on EssentialsPE and it seems to be working fine. I also used PEMapModder API tool.

**Comment by iksaku** [comment]
Fixed by 387b0484a62e47c6e470281b8bbc8e8fb7923aa3

**Comment by SuperDude36** [comment]
How is this closed? It still happens?
Fatal error: Cannot use pocketmine\nbt\tag\Float as Float because 'Float' is a special class name in phar:///Users/superserver/ServerTest/plugins/EssentialsPE_v1.1.3.phar/src/EssentialsPE/Loader.php on line 106
